primitives/npos-elections/src/lib.rs

```diff
@@ -144,9 +145,84 @@ pub type VoteWeight = u64;
 /// A type in which performing operations on vote weights are safe.
 pub type ExtendedBalance = u128;
 
-/// The score of an assignment. This can be computed from the support map via
-/// [`EvaluateSupport::evaluate`].
-pub type ElectionScore = [ExtendedBalance; 3];
+/// The score of an election. This is the main measure of an election's quality.
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo, Debug, Default)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ElectionScore {
+	/// The minimal winner, in terms of total backing stake.
+	///
+	/// This parameter should be maximized.
+	pub minimal_stake: ExtendedBalance,
+	/// The sum of the total backing of all winners.
+	///
+	/// This parameter should maximized
+	pub sum_stake: ExtendedBalance,
+	/// The sum squared of the total backing of all winners, aka. the variance.
+	///
+	/// Ths parameter should be minimized.
+	pub sum_stake_squared: ExtendedBalance,
+}
+
+impl ElectionScore {
+	fn iter(&self) -> impl Iterator<Item = ExtendedBalance> {
+		vec![self.minimal_stake, self.sum_stake, self.sum_stake_squared].into_iter()
+	}
+
+	/// Compares two sets of election scores based on desirability, returning true if `self` is
+	/// strictly `threshold` better than `other`.
+	///
+	/// Evaluation is done in a lexicographic manner, and if each element of `self` is `self *
+	/// epsilon` greater or less than `other`.
+	pub fn strict_threshold_better(&self, other: &Self, threshold: impl PerThing) -> bool {
+		match self
+			.iter()
+			.zip(other.iter())
+			.map(|(this, that)| (this.ge(&that), this.tcmp(&that, threshold.mul_ceil(that))))
+			.collect::<Vec<(bool, Ordering)>>()
+			.as_slice()
+		{
+			// epsilon better in the score[0], accept.
+			[(x, Ordering::Greater), _, _] => {
+				debug_assert!(x);
+				true
+			},
+
+			// less than epsilon better in score[0], but more than epsilon better in the second.
+			[(true, Ordering::Equal), (_, Ordering::Greater), _] => true,
+
+			// less than epsilon better in score[0, 1], but more than epsilon better in the third
+			[(true, Ordering::Equal), (true, Ordering::Equal), (_, Ordering::Less)] => true,
+
+			// anything else is not a good score.
+			_ => false,
+		}
+	}
+}
+
+impl sp_std::cmp::PartialOrd for ElectionScore {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl sp_std::cmp::Ord for ElectionScore {
+	fn cmp(&self, other: &Self) -> Ordering {
+		// we delegate this to the lexicographic cpm of vectors, and to incorporate that we want the
+		// third element to be minimized, we swap them.
+		vec![self.minimal_stake, self.sum_stake, other.sum_stake_squared].cmp(&vec![
```

Comment:
Seems a bit wasteful to create vecs here, can you instead swap them with 3-element arrays?

Reply:
I def tried that but gave up since that also caused some complains in the complier, and eventually decided to not bother too much for allocating 3 numbers, but can try again.